src/lib/Type.hs

```diff
@@ -161,15 +159,54 @@ instance HasType Expr where
               checkEq bs' bs
               resultTy' <- flip (foldr withBinder) bs $ typeCheck body
               checkEq resultTy resultTy'
-          VariantTy types -> do
+          VariantTy (NoExt types) -> do
             checkEq (length types) (length alts)
             forM_ (zip (toList types) alts) $ \(ty, (Abs bs body)) -> do
               [b] <- pure $ toList bs
               checkEq (getType b) ty
               resultTy' <- flip (foldr withBinder) bs $ typeCheck body
               checkEq resultTy resultTy'
+          VariantTy _ -> throw CompilerErr
+            "Can't pattern-match partially-known variants"
           _ -> throw TypeErr $ "Case analysis only supported on ADTs and variants, not on " ++ pprint ety
       return resultTy
+    RecordCons items record -> do
+      types <- mapM typeCheck items
+      rty <- typeCheck record
+      rest <- case rty of
+        RecordTy rest -> return rest
+        _ -> throw TypeErr $ "Can't add fields to a non-record object "
+                          <> pprint record <> " (of type " <> pprint rty <> ")"
+      return $ RecordTy $ joinExtLabeledItems types rest
+    RecordSplit types record -> do
+      mapM_ (|: TyKind) types
+      fullty <- typeCheck record
+      full <- case fullty of
+        RecordTy full -> return full
+        _ -> throw TypeErr $ "Can't split a non-record object " <> pprint record
+                          <> " (of type " <> pprint fullty <> ")"
+      diff <- labeledRowDifference full (NoExt types)
+      return $ RecordTy $ NoExt $
+        Unlabeled [ RecordTy $ NoExt types, RecordTy diff ]
```

Comment:
Why does this return an unlabeled record instead of a tuple? Wouldn't using pairs in more places allow us to get rid of this unlabeled business? Or is your plan to make all pairs record types at some point?

Reply:
No particular reason, it could very well be a tuple. But I think @dougalm mentioned that we might eventually make pairs a user-space type? Also, the variant case does actually need to be an unlabeled variant because `Either` is already a user-space type, so changing the record case wouldn't be enough.